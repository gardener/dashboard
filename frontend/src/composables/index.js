//
// SPDX-FileCopyrightText: 2023 SAP SE or an SAP affiliate company and Gardener contributors
//
// SPDX-License-Identifier: Apache-2.0
//

export { useApi } from './useApi'
export { useAuthn } from './useAuthn'
export { useLogger } from './useLogger'
export { useNav } from './useNav'
export { useTheme } from './useTheme'
export { useToken } from './useToken'
export { useUser } from './useUser'
