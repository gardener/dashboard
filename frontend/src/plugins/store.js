//
// SPDX-FileCopyrightText: 2023 SAP SE or an SAP affiliate company and Gardener contributors
//
// SPDX-License-Identifier: Apache-2.0
//

import pinia from '@/store'

export default {
  install (app) {
    app.use(pinia)
  },
}
