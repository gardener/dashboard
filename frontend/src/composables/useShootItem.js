//
// SPDX-FileCopyrightText: 2024 SAP SE or an SAP affiliate company and Gardener contributors
//
// SPDX-License-Identifier: Apache-2.0
//

import {
  computed,
  isRef,
  inject,
  provide,
} from 'vue'

import { useProjectStore } from '@/store/project'
import { useCloudProfileStore } from '@/store/cloudProfile'
import { useSeedStore } from '@/store/seed'

import { errorCodesFromArray } from '@/utils/errorCodes'

import { useShootMetadata } from './useShootMetadata'
import { useShootSpec } from './useShootSpec'
import { useShootStatus } from './useShootStatus'
import { useShootInfo } from './useShootInfo'
import { useShootAccessRestrictions } from './useShootAccessRestrictions'

const forceDeleteErrorCodes = Object.freeze([
  'ERR_CLEANUP_CLUSTER_RESOURCES',
  'ERR_CONFIGURATION_PROBLEM',
  'ERR_INFRA_DEPENDENCIES',
  'ERR_INFRA_UNAUTHENTICATED',
  'ERR_INFRA_UNAUTHORIZED',
])

export function createShootItemComposable (shootItem, options = {}) {
  const {
    projectStore = useProjectStore(),
    cloudProfileStore = useCloudProfileStore(),
    seedStore = useSeedStore(),
  } = options

  if (!isRef(shootItem)) {
    throw new TypeError('First argument `shootItem` must be a ref object')
  }

  const {
    shootMetadata,
    shootNamespace,
    shootName,
    shootProjectName,
    shootCreationTimestamp,
    shootDeletionTimestamp,
    shootGeneration,
    shootUid,
    shootAnnotations,
    getShootAnnotation,
    setShootAnnotation,
    unsetShootAnnotation,
    shootLabels,
    getShootLabel,
    setShootLabel,
    unsetShootLabel,
    shootCreatedAt,
    shootCreatedBy,
    shootConfirmationDeletion,
    isShootMarkedForDeletion,
    shootConfirmationForceDeletion,
    isShootMarkedForForceDeletion,
    isShootReconciliationDeactivated,
    isNewCluster,
    shootGardenerOperation,
    shootExpirationTimestamp,
  } = useShootMetadata(shootItem, {
    projectStore,
  })

  const {
    shootSpec,
    shootPurpose,
    isTestingCluster,
    isShootActionsDisabledForPurpose,
    isShootSettingHibernated,
    shootSecretBindingName,
    shootK8sVersion,
    shootAvailableK8sUpdates,
    shootKubernetesVersionObject,
    shootSupportedPatchAvailable,
    shootSupportedUpgradeAvailable,
    shootEnableStaticTokenKubeconfig,
    shootCloudProfileName,
    shootCloudProviderKind,
    shootWorkerGroups,
    hasShootWorkerGroups,
    shootAddons,
    shootRegion,
    shootZones,
    podsCidr,
    nodesCidr,
    servicesCidr,
    shootDomain,
    isCustomShootDomain,
    shootDnsProviders,
    shootHibernationSchedules,
    shootMaintenance,
    shootControlPlaneHighAvailabilityFailureTolerance,
    shootSeedName,
    isSeedUnreachable,
  } = useShootSpec(shootItem, {
    cloudProfileStore,
    seedStore,
  })

  const {
    shootStatus,
    isShootStatusHibernated,
    isShootLastOperationTypeDelete,
    isShootLastOperationTypeControlPlaneMigrating,
    shootLastOperationTypeControlPlaneMigrationMessage,
    shootLastOperation,
    shootLastErrors,
    shootConditions,
    shootConstraints,
    shootCredentialsRotation,
    shootReadiness,
    shootObservedGeneration,
    shootTechnicalId,
    lastMaintenance,
    isLastMaintenanceFailed,
    hibernationPossibleConstraint,
    isHibernationPossible,
    hibernationPossibleMessage,
    maintenancePreconditionSatisfiedConstraint,
    isMaintenancePreconditionSatisfied,
    maintenancePreconditionSatisfiedMessage,
    caCertificateValiditiesAcceptableConstraint,
    isCACertificateValiditiesAcceptable,
    caCertificateValiditiesAcceptableMessage,
  } = useShootStatus(shootItem)

  const {
    shootInfo,
    canLinkToSeed,
  } = useShootInfo(shootItem)

  const {
    accessRestrictionList,
  } = useShootAccessRestrictions(shootItem, {
    cloudProfileStore,
  })

  const isShootStatusHibernationProgressing = computed(() => {
    return isShootSettingHibernated.value !== isShootStatusHibernated.value
  })

  const canForceDeleteShoot = computed(() => {
    if (!shootDeletionTimestamp.value) {
      return false
    }

    const shootErrorCodes = errorCodesFromArray(shootLastErrors.value)
    return shootErrorCodes.some(item => forceDeleteErrorCodes.includes(item))
  })

  return {
    shootItem,
    /* metadata */
    shootMetadata,
    shootNamespace,
    shootName,
    shootProjectName,
    shootCreationTimestamp,
    shootDeletionTimestamp,
    shootGeneration,
    shootUid,
    shootAnnotations,
    getShootAnnotation,
    setShootAnnotation,
    unsetShootAnnotation,
    shootLabels,
    getShootLabel,
    setShootLabel,
    unsetShootLabel,
    shootCreatedAt,
    shootCreatedBy,
    shootConfirmationDeletion,
    isShootMarkedForDeletion,
    shootConfirmationForceDeletion,
    isShootMarkedForForceDeletion,
    isShootReconciliationDeactivated,
    isNewCluster,
    shootGardenerOperation,
    shootExpirationTimestamp,
    /* spec */
    shootSpec,
    shootPurpose,
    isTestingCluster,
    isShootActionsDisabledForPurpose,
    isShootSettingHibernated,
    shootSecretBindingName,
    shootK8sVersion,
    shootAvailableK8sUpdates,
    shootKubernetesVersionObject,
    shootSupportedPatchAvailable,
    shootSupportedUpgradeAvailable,
    shootEnableStaticTokenKubeconfig,
    shootCloudProfileName,
    shootCloudProviderKind,
    shootWorkerGroups,
    hasShootWorkerGroups,
    shootAddons,
    shootRegion,
    shootZones,
    podsCidr,
    nodesCidr,
    servicesCidr,
    shootDomain,
    isCustomShootDomain,
    shootDnsProviders,
    shootHibernationSchedules,
    shootMaintenance,
    shootControlPlaneHighAvailabilityFailureTolerance,
    shootSeedName,
    isSeedUnreachable,
    /* status */
    shootStatus,
    isShootStatusHibernated,
    isShootLastOperationTypeDelete,
    isShootLastOperationTypeControlPlaneMigrating,
    shootLastOperationTypeControlPlaneMigrationMessage,
    shootLastOperation,
    shootLastErrors,
    shootConditions,
    shootConstraints,
    shootCredentialsRotation,
    shootReadiness,
    shootObservedGeneration,
    shootTechnicalId,
    lastMaintenance,
    isLastMaintenanceFailed,
    hibernationPossibleConstraint,
    isHibernationPossible,
    hibernationPossibleMessage,
    maintenancePreconditionSatisfiedConstraint,
    isMaintenancePreconditionSatisfied,
    maintenancePreconditionSatisfiedMessage,
    caCertificateValiditiesAcceptableConstraint,
    isCACertificateValiditiesAcceptable,
    caCertificateValiditiesAcceptableMessage,
    /* info */
    shootInfo,
    canLinkToSeed,
    /* others */
    shootAccessRestrictions: accessRestrictionList,
    isShootStatusHibernationProgressing,
    canForceDeleteShoot,
  }
}

export function useShootItem () {
  return inject('shoot-item', null)
}

export function useProvideShootItem (shootItem, options) {
  const composable = createShootItemComposable(shootItem, options)
  provide('shoot-item', composable)
  return composable
}
