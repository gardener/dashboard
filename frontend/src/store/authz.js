//
// SPDX-FileCopyrightText: 2023 SAP SE or an SAP affiliate company and Gardener contributors
//
// SPDX-License-Identifier: Apache-2.0
//

import {
  defineStore,
  acceptHMRUpdate,
} from 'pinia'
import {
  ref,
  computed,
} from 'vue'

import { useApi } from '@/composables/useApi'

import { canI } from '@/utils'

import { useConfigStore } from './config'
import { useAuthnStore } from './authn'
import { useProjectStore } from './project'

export const useAuthzStore = defineStore('authz', () => {
  const api = useApi()
  const authnStore = useAuthnStore()
  const configStore = useConfigStore()
  const projectStore = useProjectStore()

  const spec = ref(null)
  const status = ref(null)

  const namespace = computed(() => {
    return spec.value?.namespace
  })

  const canCreateTerminals = computed(() => {
    return canI(status.value, 'create', 'dashboard.gardener.cloud', 'terminals')
  })

  const canCreateShoots = computed(() => {
    return canI(status.value, 'create', 'core.gardener.cloud', 'shoots')
  })

  const canPatchShoots = computed(() => {
    return canI(status.value, 'patch', 'core.gardener.cloud', 'shoots')
  })

  const canDeleteShoots = computed(() => {
    return canI(status.value, 'delete', 'core.gardener.cloud', 'shoots')
  })

  const canPatchShootsBinding = computed(() => {
    return canI(status.value, 'patch', 'core.gardener.cloud', 'shoots/binding')
  })

  const canGetSecrets = computed(() => {
    return canI(status.value, 'list', '', 'secrets')
  })

  const canCreateSecrets = computed(() => {
    return canI(status.value, 'create', '', 'secrets')
  })

  const canCreateShootsAdminkubeconfig = computed(() => {
    return canI(status.value, 'create', 'core.gardener.cloud', 'shoots/adminkubeconfig')
  })

  const canCreateShootsViewerkubeconfig = computed(() => {
    return canI(status.value, 'create', 'core.gardener.cloud', 'shoots/viewerkubeconfig')
  })

  const canPatchSecrets = computed(() => {
    return canI(status.value, 'patch', '', 'secrets')
  })

  const canDeleteSecrets = computed(() => {
    return canI(status.value, 'delete', '', 'secrets')
  })

  const canCreateTokenRequest = computed(() => {
    return canI(status.value, 'create', '', 'serviceaccounts/token')
  })

  const canCreateServiceAccounts = computed(() => {
    return canI(status.value, 'create', '', 'serviceaccounts')
  })

  const canPatchServiceAccounts = computed(() => {
    return canI(status.value, 'patch', '', 'serviceaccounts')
  })

  const canDeleteServiceAccounts = computed(() => {
    return canI(status.value, 'delete', '', 'serviceaccounts')
  })

  const canCreateProject = computed(() => {
    return canI(status.value, 'create', 'core.gardener.cloud', 'projects')
  })

  const canPatchProject = computed(() => {
    return canI(status.value, 'patch', 'core.gardener.cloud', 'projects', projectStore.projectName)
  })

  const canManageMembers = computed(() => {
    return canI(status.value, 'manage-members', 'core.gardener.cloud', 'projects', projectStore.projectName)
  })

  const canManageServiceAccountMembers = computed(() => {
    return canPatchProject.value || canManageMembers.value
  })

  const canDeleteProject = computed(() => {
    return canI(status.value, 'delete', 'core.gardener.cloud', 'projects', projectStore.projectName)
  })

  const canGetProjectTerminalShortcuts = computed(() => {
    return canGetSecrets.value
  })

  const canUseProjectTerminalShortcuts = computed(() => {
    return configStore.isProjectTerminalShortcutsEnabled &&
      canGetProjectTerminalShortcuts.value &&
      canCreateTerminals.value
  })

  const hasGardenTerminalAccess = computed(() => {
    return configStore.isTerminalEnabled &&
    canCreateTerminals.value &&
    canPatchServiceAccounts.value &&
    canCreateServiceAccounts.value
  })

  const hasControlPlaneTerminalAccess = computed(() => {
    return configStore.isTerminalEnabled &&
      canCreateTerminals.value &&
      authnStore.isAdmin
  })

  const hasShootTerminalAccess = computed(() => {
    return configStore.isTerminalEnabled &&
      canCreateTerminals.value
  })

  // reuse function not exported
  async function getRules (namespace) {
    const body = { namespace }
    const response = await api.getSubjectRules(body)
    status.value = response.data
  }

  async function fetchRules (namespace) {
    /**
     * The value of `spec.value?.namespace` is:
     * - undefined if no rules have been fetched yet
     * - null if only cluster scoped rules have been fetched
     * - a non-empty string if both cluster scoped rules and the rules for the namespace have been fetched
     */
    if (!namespace) {
      namespace = null
    }
    if (spec.value?.namespace !== namespace) {
      await getRules(namespace)
      this.setNamespace(namespace)
    }
  }

  function refreshRules () {
    return getRules(spec.value?.namespace)
  }

  function setNamespace (namespace) {
    spec.value = { namespace }
  }

  function $reset () {
    spec.value = null
    status.value = null
  }

  return {
    namespace,
    setNamespace,
    canCreateTerminals,
    canCreateShoots,
    canPatchShoots,
    canDeleteShoots,
    canPatchShootsBinding,
    canGetSecrets,
    canCreateSecrets,
    canCreateShootsAdminkubeconfig,
    canCreateShootsViewerkubeconfig,
    canPatchSecrets,
    canDeleteSecrets,
    canCreateTokenRequest,
    canCreateServiceAccounts,
    canPatchServiceAccounts,
    canDeleteServiceAccounts,
    canCreateProject,
    canPatchProject,
    canDeleteProject,
    canManageMembers,
    canManageServiceAccountMembers,
    canGetProjectTerminalShortcuts,
    canUseProjectTerminalShortcuts,
    hasGardenTerminalAccess,
    hasControlPlaneTerminalAccess,
    hasShootTerminalAccess,
    fetchRules,
    refreshRules,
    $reset,
  }
})

if (import.meta.hot) {
  import.meta.hot.accept(acceptHMRUpdate(useAuthzStore, import.meta.hot))
}
