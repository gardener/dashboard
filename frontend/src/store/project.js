//
// SPDX-FileCopyrightText: 2023 SAP SE or an SAP affiliate company and Gardener contributors
//
// SPDX-License-Identifier: Apache-2.0
//

import {
  defineStore,
  acceptHMRUpdate,
} from 'pinia'
import {
  ref,
  computed,
  toRef,
} from 'vue'

import { useApi } from '@/composables/useApi'

import { useAuthzStore } from './authz'
import { useAppStore } from './app'

import {
  filter,
  find,
  findIndex,
  map,
  get,
  replace,
} from '@/lodash'

export const useProjectStore = defineStore('project', () => {
  const api = useApi()
  const appStore = useAppStore()
  const authzStore = useAuthzStore()

  const list = ref(null)

  const isInitial = computed(() => {
    return list.value === null
  })

  const namespace = toRef(authzStore, 'namespace')

  const namespaces = computed(() => {
    return map(list.value, 'spec.namespace')
  })

  const projectNameMap = computed(() => {
    const projectNames = {}
    if (Array.isArray(list.value)) {
      for (const { metadata: { name }, spec: { namespace } } of list.value) {
        projectNames[namespace] = name
      }
    }
    return projectNames
  })

  const defaultNamespace = computed(() => {
    if (namespaces.value) {
      return namespaces.value.includes('garden')
        ? 'garden'
        : namespaces.value[0]
    }
    return ''
  })

  const currentNamespaces = computed(() => {
    if (namespace.value === '_all') {
      return namespaces.value
    }
    if (namespace.value) {
      return [namespace.value]
    }
    return []
  })

  const projectList = computed(() => {
    return list.value ?? []
  })

  const projectsNotMarkedForDeletion = computed(() => {
    return filter(projectList.value, project => !get(project, 'metadata.deletionTimestamp'))
  })

  const project = computed(() => {
    return find(list.value, ['spec.namespace', namespace.value])
  })

  const projectName = computed(() => {
    return get(project.value, 'metadata.name')
  })

  const projectNames = computed(() => {
    return map(list.value, 'metadata.name')
  })

  function updateList (obj) {
    const index = findIndex(list.value, ['metadata.name', obj.metadata.name])
    if (index !== -1) {
      list.value.splice(index, 1, Object.assign(list.value[index], obj))
    } else {
      list.value.push(obj)
    }
  }

  function isCurrentNamespace (namespace) {
    return currentNamespaces.value.includes(namespace)
  }

  function projectNameByNamespace (metadata) {
    const namespace = typeof metadata === 'string'
      ? metadata
      : metadata?.namespace
    return projectNameMap.value[namespace] ?? replace(namespace, /^garden-/, '')
  }

  async function fetchProjects () {
    const response = await api.getProjects()
    list.value = response.data
  }

  async function createProject (project) {
    const response = await api.createProject({ data: project })
    appStore.setSuccess('Project created')
    updateList(response.data)

    return response.data
  }

  async function patchProject (project) {
    const response = await api.patchProject({
      name: get(project, 'metadata.name', projectName.value),
      data: project,
    })
    updateList(response.data)
  }

  async function updateProject (project) {
    const response = await api.updateProject({
      name: get(project, 'metadata.name', projectName.value),
      data: project,
    })
    appStore.setSuccess('Project updated')
    updateList(response.data)
  }

  async function deleteProject (project) {
    const response = await api.deleteProject({
      name: get(project, 'metadata.name', projectName.value),
    })
    updateList(response.data)
    appStore.setSuccess('Project deleted')
    // do not remove project from store as it will stay in terminating phase for a while
  }

  async function $reset () {
    list.value = null
  }

  return {
    list,
    isInitial,
    namespace,
    namespaces,
    currentNamespaces,
    defaultNamespace,
    projectName,
    projectList,
    projectsNotMarkedForDeletion,
    project,
    projectNames,
    isCurrentNamespace,
    fetchProjects,
    createProject,
    patchProject,
    updateProject,
    deleteProject,
    projectNameByNamespace,
    $reset,
  }
})

if (import.meta.hot) {
  import.meta.hot.accept(acceptHMRUpdate(useProjectStore, import.meta.hot))
}
