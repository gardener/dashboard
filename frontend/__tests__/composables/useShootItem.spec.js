//
// SPDX-FileCopyrightText: 2023 SAP SE or an SAP affiliate company and Gardener contributors
//
// SPDX-License-Identifier: Apache-2.0
//

import {
  shallowRef,
  computed,
  reactive,
} from 'vue'
import {
  setActivePinia,
  createPinia,
} from 'pinia'

import { useAuthzStore } from '@/store/authz'
import { useShootStore } from '@/store/shoot'
import { useProjectStore } from '@/store/project'
import { useCloudProfileStore } from '@/store/cloudProfile'
import { useSeedStore } from '@/store/seed'

import { createShootItemComposable } from '@/composables/useShootItem'

import {
  set,
  cloneDeep,
  unset,
} from '@/lodash'

describe('composables', () => {
  describe('useProvideShootItem', () => {
    let shootItem
    let shootStore
    let authzStore
    let projectStore
    let cloudProfileStore
    let seedStore
    let reactiveShootItem

    function setObjectValue (object, path, value) {
      object = cloneDeep(object)
      if (value) {
        return set(object, path, value)
      }
      unset(object, path)
      return object
    }

    function setShootItem (path, value) {
      shootItem.value = setObjectValue(shootItem.value, path, value)
      shootStore.state.shoots[shootItem.value.metadata.uid] = shootItem.value
    }

    beforeEach(() => {
      setActivePinia(createPinia())
      shootItem = shallowRef({
        metadata: {
          namespace: 'garden-test',
          name: 'foo',
          uid: '1d8a140f-3e0d-4d80-a044-a2e8473c0e2d',
        },
      })

      authzStore = useAuthzStore()
      authzStore.setNamespace(shootItem.value.metadata.namespace)
      shootStore = useShootStore()
      projectStore = useProjectStore()
      cloudProfileStore = useCloudProfileStore()
      seedStore = useSeedStore()

      reactiveShootItem = reactive({
        isStaleShoot: computed(() => !shootStore.isShootActive(shootItem.value?.metadata.uid)),
        ...createShootItemComposable(shootItem, {
          projectStore,
          cloudProfileStore,
          seedStore,
        }),
      })
    })

    it('should compute isShootMarkedForDeletion correctly', () => {
      setShootItem('metadata.annotations["confirmation.gardener.cloud/deletion"]', 'True')
      expect(reactiveShootItem.isShootMarkedForDeletion).toBe(false)

      setShootItem('metadata.deletionTimestamp', '2023-01-01T20:57:01Z')
      expect(reactiveShootItem.isShootMarkedForDeletion).toBe(true)

      setShootItem('metadata.annotations["confirmation.gardener.cloud/deletion"]', 'Foo')
      expect(reactiveShootItem.isShootMarkedForDeletion).toBe(false)
    })

    it('should compute isShootMarkedForForceDeletion correctly', () => {
      setShootItem('metadata.annotations["confirmation.gardener.cloud/force-deletion"]', 'True')
      expect(reactiveShootItem.isShootMarkedForForceDeletion).toBe(false)

      setShootItem('metadata.deletionTimestamp', '2023-01-01T20:57:01Z')
      expect(reactiveShootItem.isShootMarkedForForceDeletion).toBe(true)

      setShootItem('metadata.annotations["confirmation.gardener.cloud/force-deletion"]', 'Foo')
      expect(reactiveShootItem.isShootMarkedForForceDeletion).toBe(false)
    })

    it('should compute isShootReconciliationDeactivated correctly', () => {
      setShootItem('metadata.annotations["shoot.gardener.cloud/ignore"]', 'True')
      expect(reactiveShootItem.isShootReconciliationDeactivated).toBe(true)

      setShootItem('metadata.annotations["shoot.gardener.cloud/ignore"]', 'Foo')
      expect(reactiveShootItem.isShootReconciliationDeactivated).toBe(false)
    })

    it('should compute isShootStatusHibernationProgressing correctly', () => {
      setShootItem('spec.hibernation.enabled', true)
      expect(reactiveShootItem.isShootSettingHibernated).toBe(true)
      expect(reactiveShootItem.isShootStatusHibernationProgressing).toBe(true)

      setShootItem('status.hibernated', true)
      expect(reactiveShootItem.isShootSettingHibernated).toBe(true)
      expect(reactiveShootItem.isShootStatusHibernationProgressing).toBe(false)
    })

    it('should compute isCustomShootDomain correctly', () => {
      setShootItem('spec.dns.providers', [
        { primary: false },
      ])
      expect(reactiveShootItem.isCustomShootDomain).toBe(false)

      setShootItem('spec.dns.providers', [
        { primary: false },
        { primary: true },
      ])
      expect(reactiveShootItem.isCustomShootDomain).toBe(true)
    })

    it('should compute isShootLastOperationTypeDelete correctly', () => {
      setShootItem('status.lastOperation.type', 'Delete')
      expect(reactiveShootItem.isShootLastOperationTypeDelete).toBe(true)

      setShootItem('status.lastOperation.type', 'Foo')
      expect(reactiveShootItem.isShootLastOperationTypeDelete).toBe(false)
    })

    it('should compute isShootLastOperationTypeControlPlaneMigrating correctly', () => {
      setShootItem('status.lastOperation', {
        type: 'Migrate',
        state: 'Succeeded',
      })
      expect(reactiveShootItem.isShootLastOperationTypeControlPlaneMigrating).toBe(true)

      setShootItem('status.lastOperation.type', 'Foo')
      expect(reactiveShootItem.isShootLastOperationTypeControlPlaneMigrating).toBe(false)

      setShootItem('status.lastOperation.type', 'Restore')
      expect(reactiveShootItem.isShootLastOperationTypeControlPlaneMigrating).toBe(false)

      setShootItem('status.lastOperation.state', undefined)
      expect(reactiveShootItem.isShootLastOperationTypeControlPlaneMigrating).toBe(true)
    })

    it('should compute isHibernationPossible correctly', () => {
      setShootItem('status.constraints', [{
        type: 'HibernationPossible',
        status: 'False',
      }])
      expect(reactiveShootItem.isHibernationPossible).toBe(false)

      setShootItem('status.constraints', [{
        type: 'HibernationPossible',
        status: 'True',
      }])
      expect(reactiveShootItem.isHibernationPossible).toBe(true)

      setShootItem('status.constraints', [])
      expect(reactiveShootItem.isHibernationPossible).toBe(true)
    })

    it('should compute isMaintenancePreconditionSatisfied correctly', () => {
      setShootItem('status.constraints', [{
        type: 'MaintenancePreconditionsSatisfied',
        status: 'False',
      }])
      expect(reactiveShootItem.isMaintenancePreconditionSatisfied).toBe(false)

      setShootItem('status.constraints', [{
        type: 'MaintenancePreconditionsSatisfied',
        status: 'True',
      }])
      expect(reactiveShootItem.isMaintenancePreconditionSatisfied).toBe(true)

      setShootItem('status.constraints', [])
      expect(reactiveShootItem.isMaintenancePreconditionSatisfied).toBe(true)
    })

    it('should compute isCACertificateValiditiesAcceptable correctly', () => {
      setShootItem('status.constraints', [{
        type: 'CACertificateValiditiesAcceptable',
        status: 'False',
      }])
      expect(reactiveShootItem.isCACertificateValiditiesAcceptable).toBe(false)

      setShootItem('status.constraints', [{
        type: 'CACertificateValiditiesAcceptable',
        status: 'True',
      }])
      expect(reactiveShootItem.isCACertificateValiditiesAcceptable).toBe(true)

      setShootItem('status.constraints', [])
      expect(reactiveShootItem.isCACertificateValiditiesAcceptable).toBe(true)
    })

    it('should compute isStaleShoot correctly', () => {
      authzStore.setNamespace('_all')
      shootStore.state.shootListFilters = {
        onlyShootsWithIssues: true,
        progressing: true,
      }
      setShootItem('metadata.labels["shoot.gardener.cloud/status"]', 'progressing')
      expect(reactiveShootItem.isStaleShoot).toBe(true)

      setShootItem('metadata.labels["shoot.gardener.cloud/status"]', undefined)
      expect(reactiveShootItem.isStaleShoot).toBe(false)
    })

    it('should compute canForceDeleteShoot correctly', () => {
      setShootItem('metadata.deletionTimestamp', '2023-01-01T20:57:01Z')
      expect(reactiveShootItem.canForceDeleteShoot).toBe(false)

      setShootItem('status.lastErrors', [{
        codes: ['ERR_CLEANUP_CLUSTER_RESOURCES'],
      }])
      expect(reactiveShootItem.canForceDeleteShoot).toBe(true)

      setShootItem('metadata.deletionTimestamp', undefined)
      expect(reactiveShootItem.canForceDeleteShoot).toBe(false)
    })
  })
})
