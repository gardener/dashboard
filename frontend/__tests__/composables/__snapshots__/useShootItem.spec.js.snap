// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`composables > useProvideShootItem > should compute shootCloudProviderBinding correctly 1`] = `
{
  "credentialsRef": {
    "kind": "Secret",
    "name": "aws-secret",
    "namespace": "garden-test",
  },
  "kind": "CredentialsBinding",
  "metadata": {
    "name": "aws-credentialsbinding",
    "namespace": "garden-test",
    "uid": "cb-aws-secret-uid",
  },
  "provider": {
    "type": "aws",
  },
}
`;

exports[`composables > useProvideShootItem > should compute shootCloudProviderBinding correctly 2`] = `
{
  "kind": "SecretBinding",
  "metadata": {
    "name": "aws-secretbinding",
    "namespace": "garden-test",
    "uid": "sb-aws-uid",
  },
  "provider": {
    "type": "aws",
  },
  "secretRef": {
    "name": "aws-secret",
    "namespace": "garden-test",
  },
}
`;
